### DISCUSSION TITLE
Issue setting up Drone engage

### DISCUSSION THREAD

**Discussion by Bhanupraveen** [DroneEngage/droneengage_communication#2] [General]
I am running Raspberry pi 4 and pixhawk cube mini, setup is working and i could partially get information to cloud. But if i do a restart, the connection stability is not good and parameters are coming and going. currently i am working on bench setup, with local wifi and also struggling to make the camera module work.

**Comment by Bhanupraveen** [comment]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/9f17a157-559d-4570-abc9-6f1f33e2c1ad)
This is how my camera config file as of now default,no changes done.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/f26581af-97a3-4b01-a6f6-5ca95bc80292)

**Reply by Bhanupraveen** [reply]
I resolved the telemetry streaming part, it is working fully now.

**Reply by HefnySco** [reply]
would you share please how did you solved it?

**Reply by Bhanupraveen** [reply]
I reinstalled Rasbian OS again and did again the settings , carefully . The main issue i found is , i dont know if it my board issue or the software, Telemetry 2 is not stable, telemetry 1 is quite stable. Now i did many But i dont have success with camera , I installed a Sixfab 4g hat for network.

**Reply by Bhanupraveen** [reply]
I am also configuring new autopilot to see if that is a board issue .

**Comment by HefnySco** [comment]
Just for info:
if you remove **"camera_list"** and only use:

```
"camera_start_index": 0,
"camera_end_index": 10,

```
What will happen here is that the system will scan for video drivers from /dev/video0 till /dev/video10  regardless they are valid camera or other non-camera video drivers. . And they will be all accessible from WebClient. 

using **"camera_list"**  is better of course if you know your camera drivers and want to give them name such as "front-camera", "back-camera" ...etc.

**Reply by Bhanupraveen** [reply]
And also the camera tab on webpage , icant able to click it.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/69f7b54d-499b-4b6f-bdab-3cabace2bf98)
dont know why it says, no such file

**Reply by HefnySco** [reply]
if you have the camera module up and running and de_comm.so up and running then you should be able to see the camera and video icons enabled.  Otherwise your camera module is not running or not connected to de_comm.

**Reply by HefnySco** [reply]
If you can not see the files... please try to reinstall again or manually copy the files to destination folder.

**Reply by Bhanupraveen** [reply]
Actually the problem is , camera it self not detected by board,, i am figuring out to solve this issue.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/5c9bea5b-044e-49ed-aec4-db10681b4a96)

**Reply by HefnySco** [reply]
**v4l2loopback** is a virtual camera driver ... you can use it to create a virtual camera and use ffpmeg to stream

sudo modprobe v4l2loopback devices=3
ffmpeg -f lavfi -i testsrc -pix_fmt yuv420p  -c:v rawvideo -f v4l2 /dev/video1

you can test the multiple camera feature of DE using above script. you need also to update the config file.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/eb0f64fe-c5a2-47db-85d9-0a60bfb23bf2)
able to make the camera work , need to check with de_camera

**Reply by HefnySco** [reply]
you need to enable legacy camera support ... I think so .

**Reply by Bhanupraveen** [reply]
I did enabled legacy camera support

**Comment by Bhanupraveen** [comment]
[Steps to follow to setup Drone Engage into Raspberry pi.docx](https://github.com/DroneEngage/droneengage_communication/files/14599856/Steps.to.follow.to.setup.Drone.Engage.into.Raspberry.pi.docx)
Detailed steps i followed , i compiled in to this document,need changes once the camera is also working.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/18c9462e-9535-46a3-93df-b6c12df1f19d)
can you comment if i am right or wrong? except this camera part rest all others are working perfectly.

**Reply by HefnySco** [reply]
**device_num:0**  for both cameras are wrong.
the number should be equal to /dev/videoX

so **device_num:X**   ===>    **/dev/videoX**

**Comment by Bhanupraveen** [comment]
![WhatsApp Image 2024-03-15 at 10 21 01 AM](https://github.com/DroneEngage/droneengage_communication/assets/16651096/d9e3ee4d-5c69-415c-843e-2914d5a42f0b)
The problem is , this file is not getting used, dont know why it says no such file, even though it is there

**Reply by Bhanupraveen** [reply]
the camera service is not able to run, like de_mavlink, dont know why. If that is running, camera can be seen automatically.

**Reply by Bhanupraveen** [reply]
Issue with the OS or , or de_camera is not installing ?

**Reply by HefnySco** [reply]
chmod +x de_camera

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/7f757855-d207-4f5e-a344-07ba0d36c78c)

**Reply by HefnySco** [reply]
you forgot to add .so

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/4f5ef0a1-41df-42b5-9a87-18eb5fceda0a)

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/4f132453-356d-48d3-8f53-ade616a6d4ee)

**Reply by HefnySco** [reply]
I am not sure what are you trying to do here.
please give executeable permission to the binary file of camera module.

**Reply by Bhanupraveen** [reply]
I installed camera file, but it is not running as service, like how de_mavlink is working, the camera service is not running.

**Reply by HefnySco** [reply]
are you sure there is a file... please try to run "ls" to list files and then change its permissions to +x

**Reply by HefnySco** [reply]
The error you get is that there is no file .... not bad not corrupted .... just not found.

**Reply by Bhanupraveen** [reply]
Yea ,got it ,but if you see in the folder the file exists.

**Reply by HefnySco** [reply]
Please try to delete the folder and reinstall camera module. it should work with no problems.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/074ba9e0-38b7-40fe-a139-affc7d7bd6fb)
I tried few times  deleting and installing, but please look at this, i think this might be the problem, i am running 32 bit, file is supporting 64 bit , if i am not wrong. Do check the screenshot

**Reply by Bhanupraveen** [reply]
So to verify it ,I did a fresh installation of legacy 64 but full ,and did installation of droneengage ,the camera is installed and I could able to run the ./dronecamera64.so and the client camera buttons got active. The camera file supports only arm64 installation.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/1403096f-544e-48fa-a9a1-b095b144cc47)
the camera is running normally, but dont know why there is no feed?

**Reply by HefnySco** [reply]
The camera is not working ... try to install ffplay and test th video first.

**Reply by Bhanupraveen** [reply]
Camera is working I am seeing it ,and also can save videos from it ,if I use libcamera-vid -o test.h264 it is saving video

**Reply by HefnySco** [reply]
libcamera-vid is not the legacy stack. you need to use legacy stack.... please try to validate camera using ffmpeg

**Reply by Bhanupraveen** [reply]
Ok, will try it.

**Comment by Bhanupraveen** [comment]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/c2c765fb-da00-4402-b748-c4eb2fc2ebb2)
keep getting this

**Reply by HefnySco** [reply]
if you want to make sure camera is working you need to run ffplay /dev/video0

**Comment by Bhanupraveen** [comment]
Yea ok ,still working on the video to work ,if I get result I'll post it here. Thank you

**Reply by HefnySco** [reply]
yes please do

**Comment by HefnySco** [comment]
I have downloaded the published version of de_camera and just extracted the zip file and executed directly the binary file and it worked normally.

![image](https://github.com/DroneEngage/droneengage_communication/assets/4184373/07975f3d-93f4-4b92-9bc6-8d8396779499)

**Reply by Bhanupraveen** [reply]
Can you tell me which version of OS you installed in raspberry pi ?

**Reply by HefnySco** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/4184373/186a88d9-12b5-4fe4-a6b3-900f5f31e6da)

![image](https://github.com/DroneEngage/droneengage_communication/assets/4184373/14bc235c-53da-4691-9bb4-61f9eecfbfd0)

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/59ac744c-de3d-459e-a184-bb0d70913648)

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/1218faa3-e936-44a6-88dc-3eb251081631)

**Reply by HefnySco** [reply]
please use **BullsEye**

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/f4c90d5a-c3b9-49c8-ba03-8c362aa1ae06)
is it working?

**Reply by HefnySco** [reply]
The first lines when you execute the app shows if it is working or not. but it seems to me it is working.

**Reply by Bhanupraveen** [reply]
Now it is working with ffplay, i could play the video and see on the pi, but it is not being seen on the cloud page. To verify this i also tested a usb webcam, when i click on video, it is turning green but no video on the cloud page. 
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/ee3a3c6e-7fe5-4dc3-8e94-e54de789d8f1)

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/4d59126a-4f04-4019-9a4c-65bf1fd43fcc)

**Reply by HefnySco** [reply]
Please in camera config file please set it as:

// Camera Devices Configuration
    //  "camera_start_index"        :0, 
    //  "camera_end_index"          :10, 
    "one_session_per_camera"    : true,
      "camera_list": [
        {"name": "ONE", "device_num": 0} // name should be unique across all cameras.
      ],

so one camera is detected, and let us see.
I have just tested my camera online and I was able to see myself.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/7ab9e074-290c-4bf2-b0f1-690b6147dae5)
Modified, it and i couldnt get video on the page, but when i clicked the phot symbol, images got saved in the pi from the camera.

**Reply by ievbayiro32** [reply]
How did you get the de_camera64.so to display? I keep getting can't execute binaries

**Reply by Bhanupraveen** [reply]
which os version you are using?

**Reply by ievbayiro32** [reply]
Bullseye, Do you have a command to chek though? This is a group project so I'm not to great with the OS understanding

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/7822fe98-c76d-4551-a594-c74395580c9e)

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/a90193f4-41d6-4a07-99ee-9cfcd3adc432)

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/3247ea29-3639-475b-9124-5e59e1254475)
currently i installed this on the pi.

**Reply by ievbayiro32** [reply]
![Screenshot 2024-04-03 111018](https://github.com/DroneEngage/droneengage_communication/assets/165714210/3190e16d-6bba-42ae-9de0-56b0826849da)

**Reply by ievbayiro32** [reply]
![Screenshot 2024-04-03 111049](https://github.com/DroneEngage/droneengage_communication/assets/165714210/297ff20b-528c-4ebc-80ae-ddb2bafc7b92)

**Reply by ievbayiro32** [reply]
For your new OS whats the difference? They both say bullseye

**Reply by Bhanupraveen** [reply]
use: uname -m

**Reply by ievbayiro32** [reply]
![Screenshot 2024-04-03 111557](https://github.com/DroneEngage/droneengage_communication/assets/165714210/c45c770b-4f6f-4de8-8ea6-83c51906ff9a)

**Reply by ievbayiro32** [reply]
What is yours?

**Reply by ievbayiro32** [reply]
Also something to note, I noticed you all have raspberry pi 4's I have a RPi 3B for reference if that may cause issues.

**Reply by Bhanupraveen** [reply]
you are running 32 bit version, install 64 bit version of it, as i suggested in above screenshot.It should look like this 
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/cf3fa25a-cc58-41ff-8be7-cd8818662cc1)

**Reply by Bhanupraveen** [reply]
dont worry about the board.

**Comment by Bhanupraveen** [comment]
> Please in camera config file please set it as:
> 
> // Camera Devices Configuration // "camera_start_index" :0, // "camera_end_index" :10, "one_session_per_camera" : true, "camera_list": [ {"name": "ONE", "device_num": 0} // name should be unique across all cameras. ],
> 
> so one camera is detected, and let us see. I have just tested my camera online and I was able to see myself.

Now everything is working, except seeing video in the webpage. If i click pictures, they are getting saved in the drone_engage folder, and record videos from the webpage, but its strange couldnt see the video at all, what would be the missing thing , i couldnt figure out.

**Reply by HefnySco** [reply]
open `chrome://webrtc-logs/`
the video is transmitted using WebRTC Protocol and to reach your screen from your RPI you need to turnserver, so most probably in it something related to network.

this is [script](https://github.com/DroneEngage/DroneEngage_ScriptWiki/blob/main/server_installation/prepare-airgapped-server.sh) to install a complete server on a RPI just in case you want to check turnserver installation code.

**Reply by Bhanupraveen** [reply]
![image](https://github.com/DroneEngage/droneengage_communication/assets/16651096/c7fff679-00a3-427b-b1cd-956e1bc5bd31)

**Reply by Bhanupraveen** [reply]
I have a doubt, should I change IP or leave it to 192.168.1.161? in machine IP it should be Raspi IP ?

**Reply by HefnySco** [reply]
you can... Please remember to register root.ct certificate in your browser to allow SSL connection.
pls check this video https://youtu.be/R1BedRTxuuY

**Reply by Bhanupraveen** [reply]
Browser on my local machine or Raspi?

**Reply by HefnySco** [reply]
This is a local server that can be accessed from any machine on your LAN as long as this certificate is registered on browser as a service provider and the host file contains the domain name and ip of the server.

**Reply by Bhanupraveen** [reply]
So to understand, how it works when it is on dorne, connected to a 4g dongle? if the setup is for local netwrork?

**Reply by HefnySco** [reply]
I gave you the local server so that you can make  check the video with.
It works over the internet & 4G normally. I just tested it yesterday. So if it worked locally  with you but not globally then it is a network issue.
